src/auth/auth.middleware.ts

```diff
@@ -0,0 +1,145 @@
+import { NextFunction, Request, Response, Handler } from 'express'
+import * as DAG_JOSE from 'dag-jose'
+import * as sha256 from '@stablelib/sha256'
+import * as u8a from 'uint8arrays'
+import { CARFactory, CAR } from 'cartonne'
+import { DiagnosticsLogger } from '@ceramicnetwork/common'
+import { DID } from 'dids'
+import KeyDIDResolver from 'key-did-resolver'
+import { ServiceMetrics } from '@ceramicnetwork/observability'
+import { METRIC_NAMES } from '../settings.js'
+
+export type AuthOpts = {
+  logger?: DiagnosticsLogger
+  allowedDIDs: Set<string>
+  isRelaxed: boolean
+}
+
+export const AUTH_BEARER_REGEXP = new RegExp(/Bearer (.*)/)
+const CAR_FACTORY = new CARFactory()
+CAR_FACTORY.codecs.add(DAG_JOSE)
+
+const VERIFIER = new DID({ resolver: KeyDIDResolver.getResolver() })
+
+export function parseAllowedDIDs(dids: string | undefined): Set<string> {
+  if (dids) {
+    const parts = dids.split(',')
+    return new Set(parts)
+  } else {
+    return new Set()
+  }
+}
+
+export function auth(opts: AuthOpts): Handler {
+  const hasAllowedDIDsList = opts.allowedDIDs.size > 0
+
+  /**
+   * @dev If the request has a did header, it means we have already confirmed the did
+   * is registered. If the request has no did, it means we have already
+   * confirmed the IP address making the request is on our allowlist. If the
+   * request contains a body, it means we have already verified the digest
+   * header can be trusted here.
+   * All of this logic mentioned above lives outside of this app.
+   * Notice that the absense of a did header or body bypasses any checks below
+   * this app will still work if the logice above is not in place.
+   */
+  return async function (req: Request, res: Response, next: NextFunction) {
+    const logger = opts.logger
+
+    // Use auth lambda
+    const didFromHeader = req.header('did')
+    if (didFromHeader && req.body) {
+      const digest = buildBodyDigest(req.header('Content-Type'), req.body)
+      if (req.header('digest') === digest) {
+        ServiceMetrics.count(METRIC_NAMES.AUTH_ALLOWED, 1, { did: didFromHeader })
+        return next()
+      } else {
+        logger?.verbose(`Disallowed: Auth lambda: Invalid digest`)
```

Comment:
Let's throw in a metric here too, but we probably do NOT want the didFromHeader here (because it's effectively an unbounded size list, which could lead to unbounded memory usage and an external DoS vector), so a simple count please.

Reply:
Also, let's throw in `relaxed` (=0 or =1) as a label here (and just above)  too, it makes some analyses easier.

Reply:
Looks like the `disallow` method updates a metric for the disallowed reason. Can add `relaxed` to it as well.